### ISSUE TITLE
tpm support

### ISSUE THREAD

**Issue by doubleddav** [churchers/vm-bhyve#540]
Since FreeBSD 14 release, bhyve now supports TPM passthrough and emulation, which is handy for recent versions of Windows.

Just wondering if this has or will be implemented into vm-bhyve?

Thanks

**Comment by yaroslav-gwit** [comment]
I am interested in trying this out too. Could you please point me to a man page, or a patch that shows more info on the `bhyve` TPM emulation? I couldn't find anything so far on this subject.

Once you know the raw `bhyve` flags, you can pass them in as a custom start string in `vm-bhyve`.

**Comment by nunotexbsd** [comment]
Hello,

I'm running vm-bhyve-devel with success with win11 using tpm.
Don't forget to `kldload tpm` first.

```
loader="uefi"
graphics="yes"
graphics_res="1920x1080"
xhci_mouse="yes"
cpu=4
memory=4G

# put up to 8 disks on a single ahci controller.
# without this, adding a disk pushes the following network devices onto higher slot numbers,
# which causes windows to see them as a new interface
ahci_device_limit="8"

# ideally this should be changed to virtio-net and drivers installed in the guest
# e1000 works out-of-the-box

#network0_type="e1000"
network0_type="virtio-net"
network0_switch="public"

#disk0_type="ahci-hd"
#disk0_type="virtio-blk"
#disk0_type="virtio-scsi"

disk0_type="nvme"
disk0_name="disk0.img"

#disk1_type="ahci-cd"
#disk1_dev="custom"
#disk1_name="/home/nunotex/Downloads/virtio-win-0.1.240.iso"

# windows expects the host to expose localtime by default, not UTC
utctime="no"

# sound, tpm
bhyve_options="-s 1,hda,play=/dev/dsp -l tpm,passthru,/dev/tpm0"

#bhyve_options="-s 5:0,passthru,1/0/0" # sample

# nvidia passthru
#passthru0="1/0/0"
#passthru1="1/0/1"

uuid="af2a6d8d-9c0c-11ee-9989-002b67dfc673"
network0_mac="58:9c:fc:07:8e:fd"
```

**Comment by nunotexbsd** [comment]
(...)
How I did install virtio drivers without setup from cd and how to install win11 with nvme:
https://forums.FreeBSD.org/threads/bhyve-tpm-emulation.91254/post-633892

**Comment by yaroslav-gwit** [comment]
Doesn't the `passthu` mean that you can only use the `tpm` on a single VM exclusively?

**Comment by nunotexbsd** [comment]
> Doesn't the `passthu` mean that you can only use the `tpm` on a single VM exclusively?

From https://wiki.freebsd.org/bhyve it supports passthru and emulation.
Didn't try emulation. Since its not used by OS, I'm using it passthru.

**Comment by yaroslav-gwit** [comment]
Yeah that's the thing, Wiki page mentions the emulation but the man page says only `passthu` mode is supported. Which is a problem, if you want to use the TPM on multiple VMs at the same time.

**Comment by GogoFC** [comment]
> I am interested in trying this out too. Could you please point me to a man page, or a patch that shows more info on the `bhyve` TPM emulation? I couldn't find anything so far on this subject.
> 
> Once you know the raw `bhyve` flags, you can pass them in as a custom VM start string in `vm-bhyve`.

I just use this on Bhyve and old Computers. It bypasses TPM and let's you install Win 11. https://github.com/AveYo/MediaCreationTool.bat 
I create a Windows 11 ISO file using that and it works on bhyve as well.

**Comment by Defenso-EBO** [comment]
Hi,

I'd like to work on this and eventually submit a PR. Before writing code I'd rather agree on the design and the config naming, since those become a public contract.

Context: bhyve now has a swtpm emulation backend (D46373), used as `-l tpm,swtpm,<socket>`. It's an LPC device, so it doesn't consume a PCI slot. swtpm itself needs to be running and listening on a UNIX socket before bhyve starts. I've been running this by hand and it works well for Windows 11 guests.

Proposed approach, modelled on the existing `vm::bhyve_device_rand`:

- A new `vm::bhyve_device_tpm` that appends `-l tpm,swtpm,${VM_DS_PATH}/${_name}/tpm.sock` when enabled.
- vm-bhyve owns the swtpm lifecycle: start the socket inside the run loop just before bhyve (after the `--destroy`), and clean it up on exit. No reliance on the system `swtpm` rc service, so the daemon lifetime stays tied to the guest.
- `startup-clear` on install vs `startup-state` afterwards can be inferred from whether an ISO is being booted, so it doesn't need to be exposed.

Config surface (this is the part I'd like feedback on):

- Per-guest: `tpm="yes"` in the guest `.conf` / template. Possibly `tpm_version` (defaults to 2.0, the only version bhyve supports) and `tpm_type` if passthru is ever folded in later.
- Optional global default in `system.conf`, read as a fallback when the guest doesn't specify anything.

I'd suggest splitting this into two PRs: first the per-guest support (the feature this issue is really about), then the global default separately, since the latter is more opinionated.

Does this direction and naming look right to you? Happy to adjust before starting.

Thanks

**Comment by Firstyear** [comment]
Not a project member, so my review of this is certainly not authoritative.

I think this all looks really good and would cover the majority of use cases. Something that might be worth adding in future is ability to configure the swtpm localca for ek provisioning. But certainly not needed initially.
